### PR TITLE
[XC] Allow generic types in x:DataType and x:Type

### LINKED ISSUE #20616: Support compiled bindings for generic ViewModels
Please support compiled bindings for ViewModels based on a generic type.

Sample:

- classes: `MyViewModel<T>`, `MyModel`
- in code-behind: `BindingContext = new MyViewModel<MyModel>()`

In Xaml, I want to bind
 
```
x:DataType="viewmodels:MyViewModel(models:MyModel)"
```

_Originally posted by @StephaneDelcroix in https://github.com/dotnet/maui/issues/20568#issuecomment-1946295454_

> @StephaneDelcroix, according to your offer, should I create an issue for this now, and tag-mention you? Just want to make sure before doing so that this is a real issue and I am not doing something obvious wrong with `x:DataType="viewmodels:BoViewModel(models:BoArticle)"`

>@Hottemax please do

  
There was a similar request in the Xamarin.Forms github: https://github.com/xamarin/Xamarin.Forms/issues/7161

### PR DESCRIPTION
**Pull request by simonrozsival**
### Description of Change

- Adds support for generic types in `x:DataType`
- Adds support for generic types in `x:Type` (compiled and parsed at runtime)

### Issues Fixed

Fixes #20616

### CODE REVIEW

File: src/Controls/tests/Xaml.UnitTests/Issues/Maui20616.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        public void XDataTypeCanBeGeneric([Values(false, true)] bool useCompiledXaml)
+        {
+            if (useCompiledXaml)
+            {
+                MockCompiler.Compile(typeof(Maui20616));
+            }
+            else
+            {
+                new Maui20616(useCompiledXaml: false);
+            }
```
Comment (StephaneDelcroix) [inline src/Controls/tests/Xaml.UnitTests/Issues/Maui20616.xaml.cs:53]: the test should be 
```csharp
var page = new Maui20616(useCompiledXaml);
...```

and somehow assert that the TypedBinding has the correct datatype, or at least that it worked

the MockCompiler.Compile isn't necessary, unless you need to detect that it fails, or report warnings

File: src/Controls/tests/Xaml.UnitTests/Issues/Maui20616.xaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
+             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui20616">
+    <Label Text="{Binding Value}" x:DataType="local:ViewModel20616(x:String)" />
+    <Label Text="{Binding Value}" x:DataType="{x:Type local:ViewModel20616(x:String)}" />
```
Comment (StephaneDelcroix) [inline src/Controls/tests/Xaml.UnitTests/Issues/Maui20616.xaml:7]: can you test that {x:Type} works with generics outside of the DataType context?

**Comment by simonrozsival** [comment]
@StephaneDelcroix please take another look

**Review by StephaneDelcroix — APPROVED** [review]
good when green

**Review by rmarinho — APPROVED** [review]

**Pull request merged**